File: FAMCS_algorithms_and_data_structures/graph_algorithm/0.1_building_roads/main.cpp
#include <iostream>
#include <fstream>

using namespace std;

class DSU{
    public:
    long *parent;
    long *size;

    DSU(long n){
        parent = new long[n+1];
        size = new long[n+1];
        for(int i = 1; i < n+1; i++){
            parent[i] = i;
            size[i] = 1;
        }
    }

    int FindSet(long x) {
        if(x == parent[x])
            return x;
        parent[x] = this->FindSet(parent[x]);
        return parent[x];
    }

    bool Union(long u, long v){
        u = FindSet(u);
        v = FindSet(v);
        if (u != v){
            parent[v] = u;
            size[u] += size[v];
            return true;
        }
        return false;
    }
};

int main()
{
    long n, q; // n - number of vertices, q - number of requests 

    ifstream fin("input.txt");
    ofstream fout("output.txt");
    fin >> n >> q;
    DSU temp(n);
    long u = 0, v = 0, count = n;
    for(int i = 0; i < q; i++){
        fin >> u >> v;
        if (temp.Union(u, v) && count > 1)
            count--;
        if(i == q - 1)
            fout << count;
        else
            fout << count << "\n";
    }
    return 0;
}

File: FAMCS_algorithms_and_data_structures/graph_algorithm/0.1_building_roads/main.py
class DSU:

    def __init__(self, n):
        self.size = [1 for i in range(n+1)]
        self.parent = [1 for i in range(n+1)]
    
    # for finding master
    def FindSet(self, x):
        if x == self.parent[x]:
            return x
        self.parent[x] = self.FindSet(self.parent[x])
        return self.parent[x]
    
    # for connecting bought units
    def Union(self, u, v):
        u = self.FindSet(u)
        v = self.FindSet(v)
        if u != v:
            if self.size[u] < self.size[v]:
                u, v = v, u
                self.parent[v] = u
                self.size[u] += self.size[v]
                return False
        return True
        

fin = open("input.txt", "r")

n, q = fin.readline().split()
n = int(n)
q = int(q)

count = n

temp = DSU(n)

fout = open("output.txt", "w")

for i in range(n):
    u, v = fin.readline().split()
    u = int(u)
    v = int(v)
    if count == 1:
        fout.write(str(count) + "\n")
        continue
    if temp.Union(u, v):
        count-=1
    fout.write(str(count) + "\n")